### PR TITLE
Authentication Framework

### PR DESCRIPTION
**Pull request by kylef**
This pull request proposes an RFC for adding an authentication framework to the API Blueprint language.

It supersedes:
- https://github.com/apiaryio/api-blueprint/issues/11
- https://github.com/apiaryio/api-blueprint/pull/201

### CODE REVIEW

File: draft/authentication.md [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+- [Rationale](#rationale)
+- [Backwards Compatibility](#backwards-compatibility)
+
+## Abstract
+
+This RFC proposes an authentication framework that you can use to mark
+resource groups, resources, actions or requests as authenticated.
+Along with defining a framework for creating authentication schemes.
+
+This RFC does not propose any specific authentication schemes, these are to be
```
Comment (pksunkara) [inline draft/authentication.md]: Those?

File: draft/authentication.md [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+created outside of this RFC.
+
+## Motivation
+
+Being able to express authentication in a machine readable way will allow tools
+around API Blueprint to utilising authentication.
```
Comment (pksunkara) [inline draft/authentication.md]: to utilise

File: draft/authentication.md [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+## Rationale
+
+### Authentication Scheme Type Definition
+
+Authentication schemes can be defined in an API under an "Authentication
+Schemes" heading in an API Blueprint. Inside this header may be a sub-heading
```
Comment (pksunkara) [inline draft/authentication.md]: can be defined in an API Blueprint under an "Authentication Schemes" header.

File: draft/authentication.md [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+```
+
+You may overide any properties defined in a base or named scheme.
+
+```apib
++ Authenticated (Passphrase)
+    + username: katie
+```
+
+You may also provide failure responses in anonymouse authentication schemes.
```
Comment (danielgtaylor) [inline draft/authentication.md]: What is an anonymous authentication scheme?

Also, spelling: anonymous**_e**_

File: draft/authentication.md [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+```apib
++ Request
+    + Authenticated (Basic)
+        + username: kyle
+        + password: b2952d03bda09cb5f63b0162fbbee77c
+```
+
+**NOTE**: *You may not define example responses that an authentication scheme
+may return within an anoymous authentication scheme inside a request example.
```
Comment (danielgtaylor) [inline draft/authentication.md]: Spelling: anoymous -> anonymous

**Comment by danielgtaylor** [comment]
Just the one question about anonymous authentication, otherwise LGTM! :+1:

**Comment by kylef** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;mPFG4X7Q&#x2F;289-propose-mson-based-authentication-syntax"> Propose MSON-Based Authentication syntax</a>

**Comment by kylef** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;mPFG4X7Q&#x2F;289-propose-mson-based-authentication-syntax"> Propose MSON-Based Authentication syntax</a>

File: draft/authentication.md (hunk 4 above)
Comment (kylef) [inline draft/authentication.md]: @danielgtaylor Anonymous compared to the named schemes defined in the "Authentication Scheme" section. Anonymous schemes are define in-line within other elements like resources/actions.

Does this make sense?

**Comment by rintcius** [comment]
What's the status of this proposal (ie when can we expect this feature to be available)? 

Also, since this is the relevant PR wrt authentication, maybe good to close the other PR's to avoid confusion? 
BTW https://docs.apiary.io/faq/authentication/ still refers to an outdated proposal.

**Comment by zdne** [comment]
I like it. Some qualifications are needed – for example how one can describe the schemes, or specify multiple responses – but as a draft it is solid. Going to merge it after I read the related PRs to verify this works as a framework.

**Comment by kylef** [comment]
@zdne I think multiple responses should already be clear, there is a parameter within the `Request` inside `Request Examples`.

For a more complete example, you could use the following to show both a request without authentication and a request with authentication.

``` apib
+ Response 200 (application/json)
    + Attributes
        + name: Kyle

+ Request
    + Authenticated (Basic)
        + username: kyle
        + password: b2952d03bda09cb5f63b0162fbbee77c
+ Response 200 (application/json)
    + Attributes
        + name: Kyle
        + email: `kyle@apiary.io`
```

Please let me know if it's not clear and I can amend.

File: draft/authentication.md [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+You specify that multiple authentication schemes are supported, using an
+`enum`.
+
+```apib
++ Authenticated (enum)
+    + (Basic)
+        + username: kyle
+        + password: b2952d03bda09cb5f63b0162fbbee77c
+    + (Passphrase)
+```
```
Comment (pksunkara) [inline draft/authentication.md:120]: We might want add another example here to show a simple version of this.

``` apib
+ Authenticated (enum[Basic, Passphrase])
```

File: draft/authentication.md [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+### Resource Groups
+
+The authenticated keyword may be used within a resource group to indicate that
+all their children resources, actions and example requests inside the
```
Comment (pksunkara) [inline draft/authentication.md:125]: `all of it's`

File: draft/authentication.md [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
++ Authenticated (enum)
+    + (Basic)
+        + username: kyle
+        + password: b2952d03bda09cb5f63b0162fbbee77c
+    + (Named)
+```
+
+### Resources
+
+Similiary to resource groups, a resource and it's children actions and examples
```
Comment (pksunkara) [inline draft/authentication.md:156]: example requests

File: draft/authentication.md [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+## Questions Collection [/questions]
+
++ Authenticated (Basic)
+    + username: kyle
+    + password: b2952d03bda09cb5f63b0162fbbee77c
+```
+
+### Actions
+
+You can mark a specific action as authenticated using the following:
```
Comment (pksunkara) [inline draft/authentication.md:175]: Please add mention of example requests under the action inheriting this.

**Comment by pksunkara** [comment]
``` apib
# Passphrase (Basic)
+ username: kyle
+ password: b2952d03bda09cb5f63b0162fbbee77c

+ Response 203 (application/json)

        {}
```

I am not sure where exactly you will be putting the description in the above example. I want to describe the Passphrase Authentication Scheme. Where do I do that?

**Comment by pksunkara** [comment]
@kylef My review finished.

**Pull request merged**